### ISSUE TITLE
Enhancement: [strict-boolean-expressions] options for operators/positions

### ISSUE THREAD

**Issue by That-Guy977** [typescript-eslint/typescript-eslint#8239]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/strict-boolean-expressions

### Description

`strict-boolean-expressions` should offer granular options for allowing/disallowing non-booleans in certain positions, such as the conditional position of a `if` statement or ternary, or when using specific boolean operators.

### Fail

```TypeScript
/* eslint
  "@typescript-eslint/strict-boolean-expressions": ["error", { "!": false }]
*/

!0n;
!{};

/* eslint
  "@typescript-eslint/strict-boolean-expressions": ["error", {
    "!": { "allowNumber": true, "allowString": false }
  }]
*/

!"";
// unspecified operators/options use default
!0n;
0n && true;
```

### Pass

```TypeScript
/* eslint
  "@typescript-eslint/strict-boolean-expressions": ["error", { "!": true }]
*/

!0n;
!{};

/* eslint
  "@typescript-eslint/strict-boolean-expressions": ["error", {
    "!": { "allowNumber": true, "allowString": false }
  }]
*/

!0;
// unspecified operators use default
if ("");
0 && "";
```

### Additional Info

This is motivated by a recent inquiry on the TypeScript Community Discord server, wanting `A` to be linted, in favor of `B`:
```ts
declare const x: number;

// A
<>{x && <SomeElement />}</>;

// B
<>{!!x && <SomeElement />}</>
```
Currently, with `allowNumber: false`, the rule will correctly error on `A`, but will not allow the preferred fix.

I'm not sure what format of options would be ideal for granular control while also being compatible with the current options, so the options in the example are just for demonstration.

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/2770

**Comment by That-Guy977** [comment]
I might be missing the point of that issue, isn't that more specifically targeting JSX? This is a suggestion for more granularity within the option, just motivated by the same issue
